File: src/config/loader.py
"""配置加载器模块"""
import os
import yaml
from typing import Tuple
from src.config.types import Config, DeepSeekConfig, TongyiConfig, GoodsData, GoodsInfo


# 内置 config.example.yaml 模板内容
CONFIG_EXAMPLE_TEMPLATE = """# DeepSeek API 配置（用于生成好评文案）
deepseek:
  # API 请求地址
  url: "https://api.deepseek.com/v1/chat/completions"
  # API 密钥
  apiKey: "your-deepseek-api-key-here"
  # 模型名称（如 deepseek-chat）
  model: "deepseek-chat"
  # 前调提示词（后续会拼接目标路径下的 goods.yaml 作为完整提示词）
  systemPrompt: |
    你是一位专业的电商运营专家，擅长撰写真实、客观的买家好评文案。
    请根据提供的商品信息生成买家好评文案。
    要求：
    - 字数80字左右
    - 使用较为中肯的语气
    - 禁止使用感叹号
    - 禁止使用夸张的语态
    - 禁止使用"太棒了"、"超级推荐"等夸张词汇
  # 好评文案生成次数（3-5次，提升可用率的命中）
  # 每次生成会基于相同的商品信息但可能产生不同的文案
  # 最终会保存所有生成的文案供用户选择
  reviewCount: 4

# 通义万相 API 配置（用于生成买家秀图片）
tongyi:
  # API 请求地址
  url: "https://dashscope.aliyuncs.com/api/v1/services/aigc/text2image/generation"
  # API 密钥
  apiKey: "your-tongyi-api-key-here"
  # 模型名称（如 wanx-v1 或 qwen3.5-flash）
  model: "wanx-v1"
  # 前调提示词（后续会拼接商品信息作为完整提示词）
  systemPrompt: |
    请生成一张电商买家秀模特图。
    要求：
    - 模特不出现面部和头部
    - 背景为家居室内客厅环境或户外草地环境
    - 随机生成合适的姿势
    - 图片主体为商品本身
    - 画面真实自然，符合真实买家秀风格
  # 输出图片数量（3-5张）
  imageCount: 4
"""

# 内置 goods.yaml 模板内容
GOODS_TEMPLATE = """# 商品基本信息
goods:
  # 商品名称
  name: "商品名称"
  # 品类
  category: "品类"
  # 适用性别
  gender: "男"
  # 适用年龄
  ageGroup: "青年"
  # 年份
  year: "2024"
  # 季节
  season: "春夏季"
  # 裤长（如果是裤子）
  pantsLength: ""
  # 袖长
  sleeveLength: "短袖"
  # 领型
  collarType: "圆领"
  # 穿戴方式
  wearType: "套头"
  # 特殊卖点或功能
  features:
    - "卖点1"
    - "卖点2"

# 商品图片文件列表（支持 jpg/jpeg/png）
images:
  - "front.jpg"
  - "back.jpg"
"""


def load_config(config_path: str = "config.yaml") -> Config:
    """加载项目根目录的 config.yaml 配置文件"""
    if not os.path.exists(config_path):
        raise FileNotFoundError(f"配置文件不存在: {config_path}，请创建配置文件")
    
    with open(config_path, 'r', encoding='utf-8') as f:
        data = yaml.safe_load(f)
    
    if not data:
        raise ValueError("配置文件为空")
    
    # 验证必要字段
    if 'deepseek' not in data:
        raise ValueError("配置文件缺少 deepseek 配置")
    if 'tongyi' not in data:
        raise ValueError("配置文件缺少 tongyi 配置")
    
    deepseek_data = data['deepseek']
    tongyi_data = data['tongyi']
    
    # 验证 DeepSeek 配置
    required_deepseek_fields = ['url', 'apiKey', 'model', 'systemPrompt', 'reviewCount']
    for field in required_deepseek_fields:
        if field not in deepseek_data:
            raise ValueError(f"DeepSeek 配置缺少必要字段: {field}")
    
    # 验证通义配置
    required_tongyi_fields = ['url', 'apiKey', 'model', 'systemPrompt', 'imageCount']
    for field in required_tongyi_fields:
        if field not in tongyi_data:
            raise ValueError(f"通义配置缺少必要字段: {field}")
    
    deepseek_config = DeepSeekConfig(
        url=deepseek_data['url'],
        apiKey=deepseek_data['apiKey'],
        model=deepseek_data['model'],
        systemPrompt=deepseek_data['systemPrompt'],
        reviewCount=deepseek_data['reviewCount']
    )
    
    tongyi_config = TongyiConfig(
        url=tongyi_data['url'],
        apiKey=tongyi_data['apiKey'],
        model=tongyi_data['model'],
        systemPrompt=tongyi_data['systemPrompt'],
        imageCount=tongyi_data['imageCount']
    )
    
    return Config(deepseek=deepseek_config, tongyi=tongyi_config)


def load_goods_data(target_dir: str) -> GoodsData:
    """加载目标路径下的 goods.yaml 商品信息"""
    goods_yaml_path = os.path.join(target_dir, "goods.yaml")
    
    if not os.path.exists(goods_yaml_path):
        raise FileNotFoundError(f"商品配置文件不存在: {goods_yaml_path}，请在目标路径下创建 goods.yaml")
    
    with open(goods_yaml_path, 'r', encoding='utf-8') as f:
        data = yaml.safe_load(f)
    
    if not data:
        raise ValueError("商品配置文件为空")
    
    if 'goods' not in data:
        raise ValueError("商品配置文件缺少 goods 字段")
    
    goods_data = data['goods']
    images = data.get('images', [])
    
    # 验证商品必要字段
    required_goods_fields = ['name', 'category', 'gender', 'ageGroup', 'year', 
                             'season', 'sleeveLength', 'collarType', 'wearType', 'features']
    for field in required_goods_fields:
        if field not in goods_data:
            raise ValueError(f"商品信息缺少必要字段: {field}")
    
    goods_info = GoodsInfo(
        name=goods_data['name'],
        category=goods_data['category'],
        gender=goods_data['gender'],
        ageGroup=goods_data['ageGroup'],
        year=goods_data['year'],
        season=goods_data['season'],
        pantsLength=goods_data.get('pantsLength', ''),
        sleeveLength=goods_data['sleeveLength'],
        collarType=goods_data['collarType'],
        wearType=goods_data['wearType'],
        features=goods_data['features']
    )
    
    return GoodsData(goods=goods_info, images=images)


def load_all_config(target_dir: str) -> Tuple[Config, GoodsData]:
    """加载所有配置"""
    config = load_config()
    goods_data = load_goods_data(target_dir)
    return config, goods_data


def check_and_generate_config() -> bool:
    """
    检查并生成配置文件
    
    Returns:
        bool: config.yaml 是否已存在
    """
    config_path = "config.yaml"
    example_path = "config.example.yaml"
    
    if os.path.exists(config_path):
        return True
    
    # 检查 config.example.yaml 是否存在
    if not os.path.exists(example_path):
        # 生成模板文件
        with open(example_path, 'w', encoding='utf-8') as f:
            f.write(CONFIG_EXAMPLE_TEMPLATE)
    
    return False


def generate_goods_yaml_template(target_dir: str) -> str:
    """
    生成 goods.yaml 模板到目标目录
    
    Args:
        target_dir: 目标目录路径
    
    Returns:
        str: 生成的模板文件路径
    """
    goods_yaml_path = os.path.join(target_dir, "goods.yaml")
    
    # 如果已存在则不覆盖
    if os.path.exists(goods_yaml_path):
        return goods_yaml_path
    
    with open(goods_yaml_path, 'w', encoding='utf-8') as f:
        f.write(GOODS_TEMPLATE)
    
    return goods_yaml_path

File: src/config/loader.ts
/**
 * 配置文件加载器
 * 负责加载和解析项目根目录的 config.yaml 和目标路径下的 goods.yaml
 */

import { readFileSync, existsSync } from "fs";
import { resolve, join } from "path";
import YAML from "yaml";
import type { Config, GoodsInfo } from "./types";

/**
 * 错误类：配置文件相关错误
 */
export class ConfigError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ConfigError";
  }
}

/**
 * 加载项目根目录的 config.yaml
 * @param projectRoot - 项目根目录路径
 * @returns 解析后的配置对象
 */
export function loadConfig(projectRoot: string): Config {
  const configPath = resolve(projectRoot, "config.yaml");

  if (!existsSync(configPath)) {
    throw new ConfigError(
      `配置文件不存在: ${configPath}\n请在项目根目录创建 config.yaml 文件`
    );
  }

  try {
    const configContent = readFileSync(configPath, "utf-8");
    const config = YAML.parse(configContent) as Config;

    // 验证配置完整性
    validateConfig(config);

    return config;
  } catch (error) {
    if (error instanceof ConfigError) {
      throw error;
    }
    throw new ConfigError(
      `配置文件解析失败: ${(error as Error).message}`
    );
  }
}

/**
 * 加载目标路径下的 goods.yaml
 * @param targetDir - 目标目录路径
 * @returns 解析后的商品信息对象
 */
export function loadGoodsInfo(targetDir: string): GoodsInfo {
  const goodsPath = join(targetDir, "goods.yaml");

  if (!existsSync(goodsPath)) {
    throw new ConfigError(
      `商品配置文件不存在: ${goodsPath}\n请在目标路径下创建 goods.yaml 文件`
    );
  }

  try {
    const goodsContent = readFileSync(goodsPath, "utf-8");
    const goodsInfo = YAML.parse(goodsContent) as GoodsInfo;

    // 验证商品信息完整性
    validateGoodsInfo(goodsInfo);

    return goodsInfo;
  } catch (error) {
    if (error instanceof ConfigError) {
      throw error;
    }
    throw new ConfigError(
      `商品配置文件解析失败: ${(error as Error).message}`
    );
  }
}

/**
 * 验证配置文件的必要字段
 * @param config - 配置对象
 */
function validateConfig(config: Config): void {
  // 验证 DeepSeek 配置
  if (!config.deepseek) {
    throw new ConfigError("配置文件中缺少 deepseek 配置");
  }
  if (!config.deepseek.url) {
    throw new ConfigError("deepseek.url 不能为空");
  }
  if (!config.deepseek.apiKey) {
    throw new ConfigError("deepseek.apiKey 不能为空");
  }
  if (!config.deepseek.systemPrompt) {
    throw new ConfigError("deepseek.systemPrompt 不能为空");
  }
  if (typeof config.deepseek.reviewCount !== "number" || config.deepseek.reviewCount < 1) {
    throw new ConfigError("deepseek.reviewCount 必须是大于0的数字");
  }

  // 验证通义万相配置
  if (!config.tongyi) {
    throw new ConfigError("配置文件中缺少 tongyi 配置");
  }
  if (!config.tongyi.url) {
    throw new ConfigError("tongyi.url 不能为空");
  }
  if (!config.tongyi.apiKey) {
    throw new ConfigError("tongyi.apiKey 不能为空");
  }
  if (!config.tongyi.systemPrompt) {
    throw new ConfigError("tongyi.systemPrompt 不能为空");
  }
  if (typeof config.tongyi.imageCount !== "number" || config.tongyi.imageCount < 1) {
    throw new ConfigError("tongyi.imageCount 必须是大于0的数字");
  }
}

/**
 * 验证商品信息的必要字段
 * @param goodsInfo - 商品信息对象
 */
function validateGoodsInfo(goodsInfo: GoodsInfo): void {
  if (!goodsInfo.goods) {
    throw new ConfigError("商品配置文件中缺少 goods 字段");
  }

  const { goods } = goodsInfo;

  // 验证商品基本信息
  if (!goods.name) {
    throw new ConfigError("商品名称 (goods.name) 不能为空");
  }
  if (!goods.category) {
    throw new ConfigError("商品品类 (goods.category) 不能为空");
  }
  if (!goods.gender) {
    throw new ConfigError("适用性别 (goods.gender) 不能为空");
  }
  if (!goods.ageGroup) {
    throw new ConfigError("适用年龄 (goods.ageGroup) 不能为空");
  }
  if (!goods.year) {
    throw new ConfigError("年份 (goods.year) 不能为空");
  }
  if (!goods.season) {
    throw new ConfigError("季节 (goods.season) 不能为空");
  }
  if (!goods.sleeveLength) {
    throw new ConfigError("袖长 (goods.sleeveLength) 不能为空");
  }
  if (!goods.collarType) {
    throw new ConfigError("领型 (goods.collarType) 不能为空");
  }
  if (!goods.wearType) {
    throw new ConfigError("穿戴方式 (goods.wearType) 不能为空");
  }
  if (!goods.features || !Array.isArray(goods.features) || goods.features.length === 0) {
    throw new ConfigError("商品卖点 (goods.features) 不能为空");
  }

  // 验证图片列表
  if (!goodsInfo.images || !Array.isArray(goodsInfo.images) || goodsInfo.images.length === 0) {
    throw new ConfigError("商品图片列表 (images) 不能为空");
  }
}